mods/default/workbench.lua

```diff
@@ -0,0 +1,234 @@
+	meta:set_string("formspec", get_workbench_formspec(pos))
+end
+
+local function workbench_update_help(pos, type, string)
+	local meta = minetest.get_meta(pos)
+	meta:set_string("formspec", get_workbench_formspec(pos, string))
+	meta:set_string("error", type)
+end
+
+--
+-- Node definition
+--
+
+minetest.register_node("default:workbench", {
+	description = "Workbench",
+	tiles = {"default_workbench_top.png", "default_wood.png", "default_workbench_sides.png",
+		"default_workbench_sides.png", "default_workbench_sides.png", "default_workbench_sides.png"},
+	groups = {choppy = 3, oddly_breakable_by_hand = 2, flammable = 3},
+	sounds = default.node_sound_wood_defaults(),
+
+	drawtype = "nodebox",
+	paramtype = "light",
+	node_box = {
+		type = "fixed",
+		fixed = {
+			{-0.4375, 0.1875, -0.4375, 0.4375, 0.5, 0.4375}, -- Top
+			{-0.5, 0.25, -0.5, -0.4375, 0.5, 0.5}, -- TopSide1
+			{0.4375, 0.25, -0.5, 0.5, 0.5, 0.5}, -- TopSide2
+			{-0.5, 0.25, 0.4375, 0.5, 0.5, 0.5}, -- TopSide3
+			{-0.5, 0.25, -0.5, 0.5, 0.5, -0.4375}, -- TopSide4
+			{-0.4375, -0.5, 0.25, -0.25, 0.5, 0.4375}, -- Leg1
+			{-0.4375, -0.5, -0.4375, -0.25, 0.5, -0.25}, -- Leg2
+			{0.25, -0.5, 0.25, 0.4375, 0.5, 0.4375}, -- Leg3
+			{0.25, -0.5, -0.4375, 0.4375, 0.5, -0.25}, -- Leg4
+		}
+	},
+
+	on_construct = function(pos)
+		local meta = minetest.get_meta(pos)
+		meta:set_string("formspec", get_workbench_formspec(pos))
+		local inv = meta:get_inventory()
+		inv:set_size("input", 1)
+		inv:set_size("nametag", 1)
+		inv:set_size("output", 1)
+	end,
+	can_dig = function(pos, player)
+		local inv = minetest.get_meta(pos):get_inventory()
+
+		if inv:is_empty("input") and inv:is_empty("nametag") and inv:is_empty("output") then
+			return true
+		else
+			return false
+		end
+	end,
+	on_blast = function(pos)
+		local inv = minetest.get_meta(pos):get_inventory()
+
+		local drops = {
+			inv:get_list("input")[1],
+			inv:get_list("nametag")[1],
+			inv:get_list("output")[1],
+			"default:workbench",
+		}
+		minetest.remove_node(pos)
+		return drops
+	end,
+
+	allow_metadata_inventory_put = function(pos, listname, index, stack)
+		if not stack:is_known() then
+			return 0
+		end
+
+		if listname == "nametag" then
+			if stack:get_name() ~= "default:nametag" then
+				return 0
+			else
+				return stack:get_count()
+			end
+		elseif listname == "output" then
+			return 0
+		elseif listname == "input" then
+			if minetest.registered_items[stack:get_name()].groups.renameable == 0 then
+				return 0
+			end
+		end
+
+		return 1
+	end,
+	allow_metadata_inventory_move = function(pos, from_list, from_index, to_list, to_index, count, player)
+		local inv = minetest.get_meta(pos):get_inventory()
+		local stack = inv:get_stack(from_list, from_index)
+
+		if to_list == "nametag" then
+			if stack:get_name() ~= "default:nametag" then
+				return 0
+			end
+		elseif to_list == "input" then
+			if minetest.registered_items[stack:get_name()].groups.renameable == 0 then
```

Comment:
Also, even if an item becomes undefined or is changed to `renameable = 0` while it is sitting in the workbench it will not crash the game (or did I misunderstand your question?).

Checks for unknown or renameable items are only made before the item is inserted or renamed (latter will soon be pushed). So, no, items can't be trapped in the slots.